wordpress-mobile/WordPress-iOS
# Share Extension: Allows multiple media items to be uploaded

## Discussion
This PR changes the way we are sharing media from the existing `SLComposeServiceViewController`-based Share VC. It is core to allowing us to leverage the new app extension designs (e.g. #8137, #8138). There are a few things to note here:

**1.** The biggest change is that there is a new, separate core data stack (`SharedCoreDataStack.swift`) that persists a single managed object (`UploadOperation.swift`) to the shared container. This CD instance the fundamental communication mechanism between app extensions and WPiOS. Each "UploadOp" represents individual media items as well as the post itself. For example, if you share a post with two attached images, there will be 3 UploadOps persisted to the store — 2 records for the images, and 1 record for the post. As each upload is processed, the corresponding upload op is updated with a status, (Pending, In Progress, Complete, or Error). 

**2.** You will undoubtedly notice that we have to slightly duplicate logic between `ShareViewController.swift` and `ShareExtensionSessionManager.swift` ... specifically the completion handlers for `NSURLSession` background sessions. Unfortunately there is no way around this. It is non-deterministic as to whether the share extension or WPiOS will get the call (9 times out of 10, WPiOS does like you would expect).

### General Workflow
1. User selects 1 or more images to share in the host app.
2. `ShareExtractor` parses the context extension and extracts the images and/or text
3. User is presented with `ShareViewController`
4. If user selects "Post" on dialog:
    - The selected media is encoded and saved into the shared container
    - Upload ops are created in CD for each piece of media and the post
    - A NSURLSession background session is kicked off to upload the media
5. Server response is executed:
    - **Scenario 1:** The media upload happens quickly and the NSURLSession background session completion handler runs inside the share extension (`ShareViewController`).
    - **Scenario 2 :** The NSURLSession background session completion handler runs inside WPiOS (`ShareExtensionSessionManager`).
6. Based on the response in step 5, the corresponding upload ops are updated in the shared CD instance. 
7. The local files media files in the shared container are removed IF the server response was 👍 (not deleted if it wasn't)
8. Using the post upload op, a new post is uploaded to the server containing the remote media.
9. New post response is received and saved to the CD instance.

### Things to note:
* Because app extensions are notoriously problematic, we are taking a conservative approach and only allowing for a **max of 2 images** to be shared. We can increase this limit in a subsequent PR when we release the new custom UI _and_ see if there are any crashers reported when 9.1 is pushed.

* Keep in mind that `ShareViewController`'s days are numbered, but we will transfer a lot of the logic into the new UI classes.

* There is no retry (yet)

Fixes #8179
Ref: #8265 and #8263 

## Testing

### App extension fun
A few things to think about when you are testing:

* Xcode’s debugger prevents the system from suspending the process being debugged.  So, if you run your extension from Xcode, or you attach to its process some time after launch, the process will continue to execute in situations where the system would otherwise have suspended it.  
* Be sure to test on a device, as well as the simulator. The simulator is much less likely to terminate an extension.
* Try to test on a variety of devices iPhone 5s, iPhone X, etc.
* If the user terminates the app by force-quiting from the app switcher, the system will cancel all of the session’s background transfers, and won’t attempt to relaunch the app.

### What to test
The main thing we want to test are all of the sharing flows with media. Example flow:

1. Build and run the app on a device (NOT connected to XCode)
2. Open the photos app
3. Select 1 or 2 photos and share using the debug build
4. Hopefully see a post on your site within 10 Seconds to 3 minutes depending on connection.

Try to test different scenarios/edge cases:
* Kill the network connectivity while the background sesson is running and then turn it on again (it should eventually resume but may take a few minutes).
* Multitasking on iPad
* Share images from other apps
* Share from the host app and then quickly open WPiOS (and leave it open)
* Try a really slow connection
* Try sharing an image and then after the share UI is dismissed, quickly share another image (to create another bg session)

## Thank you!

@SergioEstevao, @frosty, @kwonye I would appreciate it if you could take a look at this! Please let me know if you have any questions!
